src/components/modules/caret.ts

```diff
@@ -552,6 +540,147 @@ export default class Caret extends Module {
     selection.addRange(newRange);
   }
 
+  /**
+   * Detect next position from current position
+   */
+  private detectNext(): {
+    nextBlock: Block;
+    nextInput?: HTMLElement;
+    downOffset: number;
+} | false {
+    const { BlockManager, Tools } = this.Editor;
+
+    const nextInput = BlockManager.currentBlock.nextInput;
+    let nextBlock = BlockManager.nextContentfulBlock;
+
+    if (!nextBlock && !nextInput) {
+      /**
+       * If there is a last block and it is default, do nothing
+       * This code allows to exit from the last non-initial tool:
+       * https://github.com/codex-team/editor.js/issues/1103
+       */
+      if (Tools.isDefault(BlockManager.currentBlock.tool)) {
+        return false;
+      }
+
+      /**
+       * If there is no nextBlock, but currentBlock is not default,
+       * insert new default block at the end and navigate to it
+       */
+      nextBlock = BlockManager.insertAtEnd();
+    }
+
+    const currentBoundingClientRect = Selection.get().getRangeAt(0)
+      .getBoundingClientRect();
+
+    const range = new Range();
+    const root = nextInput ?? nextBlock.firstInput;
+    const treeWalker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT);
+
+    let node = treeWalker.firstChild();
+    let offset = 0;
+    let prevBoundingClientRect: DOMRect | undefined;
+
+    while (node) {
+      if (!(node instanceof Text)) {
+        throw new Error('Unexpected node type');
+      }
+
+      for (let index = 0; index < node.length; index++) {
+        range.setStart(node, index);
+        range.setEnd(node, index + 1);
+
+        const boundingClientRect = range.getBoundingClientRect();
+
+        if (
+          prevBoundingClientRect &&
+          Math.abs(currentBoundingClientRect.x - prevBoundingClientRect.x) < Math.abs(currentBoundingClientRect.x - boundingClientRect.x)
+        ) {
+          return {
+            nextBlock,
+            nextInput,
+            downOffset: offset - 1,
+          };
+        }
+
+        offset++;
+        prevBoundingClientRect = boundingClientRect;
+      }
+
+      node = treeWalker.nextNode();
+    }
+
+    return {
+      nextBlock,
+      nextInput,
+      downOffset: root.textContent.length,
+    };
+  }
+
+  /**
+   * Detect previous position from current position
+   */
+  private detectPrevious(): {
+    previousBlock: Block;
+    previousInput?: HTMLElement;
+    upOffset: number;
+} | false {
+    const { BlockManager } = this.Editor;
```

Comment:
looks like there are many duplicated lines with a small difference. Maybe we can move it to some utility-method with the 'direction' parameter or something like that?

Similar methods: `detectNext`, `detectPrevious` and `isNextLineExisted`, `isPreviousLineExisted`

Reply:
OK!
I'll fix it. 

```
$ jscpd -r consoleFull src/components/modules/caret.ts 
Clone found (typescript):
 - src/components/modules/caret.ts [653:60 - 664:26] (11 lines, 102 tokens)
   src/components/modules/caret.ts [589:55 - 600:22]

Clone found (typescript):
 - src/components/modules/caret.ts [740:39 - 753:38] (13 lines, 119 tokens)
   src/components/modules/caret.ts [582:52 - 596:33]

Clone found (typescript):
 - src/components/modules/caret.ts [767:32 - 776:36] (9 lines, 99 tokens)
   src/components/modules/caret.ts [731:28 - 740:37]

Clone found (typescript):
 - src/components/modules/caret.ts [776:38 - 789:31] (13 lines, 123 tokens)
   src/components/modules/caret.ts [646:52 - 660:33]

Clone found (typescript):
 - src/components/modules/caret.ts [653:60 - 664:26] (11 lines, 102 tokens)
   src/components/modules/caret.ts [589:55 - 600:22]

 653 │ 589 │ ) {                                                                                                                                       
 654 │ 590 │         range.setStart(node, index);                                                                                                      
 655 │ 591 │         range.setEnd(node, index + 1);                                                                                                    
 656 │ 592 │                                                                                                                                           
 657 │ 593 │         const boundingClientRect = range.getBoundingClientRect();                                                                         
 658 │ 594 │                                                                                                                                           
 659 │ 595 │         if (                                                                                                                              
 660 │ 596 │           prevBoundingClientRect &&                                                                                                       
 661 │ 597 │           Math.abs(currentBoundingClientRect.x - prevBoundingClientRect.x) < Math.abs(currentBoundingClientRect.x - boundingClientRect.x) 
 662 │ 598 │         ) {                                                                                                                               
 663 │ 599 │           return {                                                                                                                        
 664 │ 600 │             previousBlock                                                                                                                 

Clone found (typescript):
 - src/components/modules/caret.ts [740:39 - 753:38] (13 lines, 119 tokens)
   src/components/modules/caret.ts [582:52 - 596:33]

 740 │ 582 │ ;                                                                 
 741 │ 583 │                                                                   
 742 │ 584 │     while (node) {                                                
 743 │ 585 │       if (!(node instanceof Text)) {                              
 744 │ 586 │         throw new Error('Unexpected node type');                  
 745 │ 587 │       }                                                           
 746 │ 588 │                                                                   
 747 │ 589 │       for (let index = 0; index < node.length; index++) {         
 748 │ 590 │         range.setStart(node, index);                              
 749 │ 591 │         range.setEnd(node, index + 1);                            
 750 │ 592 │                                                                   
 751 │ 593 │         const boundingClientRect = range.getBoundingClientRect(); 
 752 │ 594 │                                                                   
 753 │ 595 │         if (currentBoundingClientRect                             

Clone found (typescript):
 - src/components/modules/caret.ts [767:32 - 776:36] (9 lines, 99 tokens)
   src/components/modules/caret.ts [731:28 - 740:37]

 767 │ 731 │ (): boolean {                                                                                                   
 768 │ 732 │     const { BlockManager } = this.Editor;                                                                       
 769 │ 733 │                                                                                                                 
 770 │ 734 │     const currentBoundingClientRect = Selection.get().getRangeAt(0)                                             
 771 │ 735 │       .getBoundingClientRect();                                                                                 
 772 │ 736 │                                                                                                                 
 773 │ 737 │     const range = new Range();                                                                                  
 774 │ 738 │     const treeWalker = document.createTreeWalker(BlockManager.currentBlock.currentInput, NodeFilter.SHOW_TEXT); 
 775 │ 739 │                                                                                                                 
 776 │ 740 │     let node = treeWalker.lastChild                                                                             

Clone found (typescript):
 - src/components/modules/caret.ts [776:38 - 789:31] (13 lines, 123 tokens)
   src/components/modules/caret.ts [646:52 - 660:33]

 776 │ 646 │ ;                                                                 
 777 │ 647 │                                                                   
 778 │ 648 │     while (node) {                                                
 779 │ 649 │       if (!(node instanceof Text)) {                              
 780 │ 650 │         throw new Error('Unexpected node type');                  
 781 │ 651 │       }                                                           
 782 │ 652 │                                                                   
 783 │ 653 │       for (let index = node.length - 1; index >= 0; index--) {    
 784 │ 654 │         range.setStart(node, index);                              
 785 │ 655 │         range.setEnd(node, index + 1);                            
 786 │ 656 │                                                                   
 787 │ 657 │         const boundingClientRect = range.getBoundingClientRect(); 
 788 │ 658 │                                                                   
 789 │ 659 │         if (boundingClientRect                                    

Found 4 clones.
Detection time:: 171.799ms
```

Reply:
Fixed duplications.